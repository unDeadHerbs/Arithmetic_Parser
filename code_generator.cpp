#include "code_generator.hpp"
#include <stdlib.h>
#include <sys/mman.h>
#include <numeric>
using std::accumulate;
using namespace std::string_literals;

namespace {
#include <iostream>
#include <string>
void print_int(int x) { std::cout << x; }
void print_str(char* s) { std::cout << s; }
void print_bool(int b) { std::cout << (b ? "true" : "false"); }
int read_int() {
	int x;
	std::cin >> x;
	return x;
}
}  // namespace

void program::operator()() {
	auto prog =
	    (char*)mmap(NULL /*we don't care where we point to.*/,
	                buffer.size() /*size of buffer*/,
	                PROT_EXEC | PROT_READ | PROT_WRITE /*permissions*/,
	                MAP_PRIVATE | MAP_ANONYMOUS /**/,
	                -1 /*ignored as not a file*/, 0 /*ignored as not a file*/);
	if (!prog) throw "Failed to alocate space"s;
	int end(0);
	for (auto c : buffer) prog[end++] = c;
	((int (*)())prog)();
	munmap(prog, buffer.size());
}
#define LITTLEENDIAN4(x)                                                       \
	program({(char)(((x) >> 0x08 * 0) & 0xFF), (char)(((x) >> 0x08 * 1) & 0xFF), \
	         (char)(((x) >> 0x08 * 2) & 0xFF),                                   \
	         (char)(((x) >> 0x08 * 3) & 0xFF)})
#define LITTLEENDIAN8(x)                                                       \
	program({(char)(((x) >> 0x08 * 0) & 0xFF), (char)(((x) >> 0x08 * 1) & 0xFF), \
	         (char)(((x) >> 0x08 * 2) & 0xFF), (char)(((x) >> 0x08 * 3) & 0xFF), \
	         (char)(((x) >> 0x08 * 4) & 0xFF), (char)(((x) >> 0x08 * 5) & 0xFF), \
	         (char)(((x) >> 0x08 * 6) & 0xFF),                                   \
	         (char)(((x) >> 0x08 * 7) & 0xFF)})
#define _MOVI_EAX program({((char)(0xB8 + 0x0))})

#define ADD_EAX_EDX program({((char)(0x01 + 0x00)), ((char)(0xd0))})
#define SUB_EAX_EDX program({((char)(0x29 + 0x00)), ((char)(0xd0))})
#define MUL_EAX_EDX program({((char)(0x0F)), ((char)(0xAF)), ((char)(0xC2))})
#define DIV_EDAX_ECX program({((char)(0xF7)), ((char)(0xF9))})

#define PUSH_EAX program({((char)(0x50 + 0x00))})
#define PUSH_ECX program({((char)(0x50 + 0x01))})
#define PUSH_EDX program({((char)(0x50 + 0x02))})
#define POP_EAX program({((char)(0x58 + 0x00))})
#define POP_ECX program({((char)(0x58 + 0x01))})
#define POP_EDX program({((char)(0x58 + 0x02))})
#define POP_RDI program({((char)(0x58 + 0x07))})
#define MOV_RDI_RAX program({((char)0x89), ((char)0xc7)})
#define MOD_EDAX_ECX DIV_EDAX_ECX + PUSH_EDX + POP_EAX
#define LOADI(x) _MOVI_EAX + LITTLEENDIAN4((int)(x))
#define ZERO_EAX LOADI(0)
#define LOADI_S(x) LOADI(strtol(x.c_str(), NULL, 10))
#define _MOVI_RSI program({((char)(0x48)), ((char)(0xbe))})
#define _MOVI_RDI program({((char)(0x48)), ((char)(0xbf))})
#define _MOVI_RAX program({((char)(0x48)), ((char)(0xc8))})
#define LOADLL_RDI(x) _MOVI_RDI + LITTLEENDIAN8((x))
#define LOADLL_RAX(x) _MOVI_RAX + LITTLEENDIAN8((x))
#define LOADLL_STR(x) _MOVI_RDI + LITTLEENDIAN8((x))
#define LOADLL_RSI(x) _MOVI_RSI + LITTLEENDIAN8((x))
#define CALL_RSI program({((char)(0xff)), ((char)(0xd6))})
#define PRINT_EAX_I \
	MOV_RDI_RAX + LOADLL_RSI((long long int)print_int) + CALL_RSI
#define PRINT_EAX_BOOL \
	MOV_RDI_RAX + LOADLL_RSI((long long int)print_bool) + CALL_RSI
#define PRINT_STR LOADLL_RSI((long long int)print_str) + CALL_RSI
#define READ_EAX LOADLL_RSI((long long int)read_int) + CALL_RSI
#define MOV_EAX_STACK(d) \
	program({((char)0x8b), ((char)0x45), ((char)((-4 * (d + 1)) & 0xff))})
#define MOV_STACK_EAX(d) \
	program({((char)0x89), ((char)0x45), ((char)((-4 * (d + 1)) & 0xff))})
// mov eax,[ebp-4*d]
#define _BOOL_AL_TO_EAX \
	program({((char)0xc0), ((char)0x66), ((char)0x98), ((char)0x98)})
#define CMP_EAX_ECX program({((char)0x39), ((char)0xc8)})
#define BOOL_IS_LESS program({((char)0x0f), ((char)0x9c)}) + _BOOL_AL_TO_EAX
#define BOOL_EAX_LESS_ECX CMP_EAX_ECX + BOOL_IS_LESS
#define BOOL_IS_LESS_EQ program({((char)0x0f), ((char)0x9e)}) + _BOOL_AL_TO_EAX
#define BOOL_IS_EQ program({((char)0x0f), ((char)0x94)}) + _BOOL_AL_TO_EAX
#define BOOL_IS_GREATER_EQ \
	program({((char)0x0f), ((char)0x9d)}) + _BOOL_AL_TO_EAX
#define BOOL_IS_GREATER program({((char)0x0f), ((char)0x9f)}) + _BOOL_AL_TO_EAX
#define BOOL_IS_NE program({((char)0x0f), ((char)0x95)}) + _BOOL_AL_TO_EAX
#define BOOL_EAX_OR_ECX program({((char)0x09), ((char)0xc8)})
#define BOOL_EAX_AND_ECX program({((char)0x21), ((char)0xc8)})
//#define BOOL_NOT program({((char)0xf7), ((char)0xd0)})
#define BOOL_NOT PUSH_EAX + LOADI(1) + POP_EDX + SUB_EAX_EDX
// TODO: make that not command much more effecent

program generateI(Code_Tree ct, std::shared_ptr<Env> e) {
	// TODO: Those throws should get the TOKEN's name.
	if (ct.name == "numeric_literal") {
		if (ct.t->id == INT) return LOADI_S(ct.t->text);
		throw "Unsupported Literal Type"s;
	}
	if (ct.name == "bool_literal") {
		if (ct.t->text == "true") return LOADI(1);
		if (ct.t->text == "false") return LOADI(0);
		throw "Unsupported bool literal '"s + ct.t->text + "'";
	}
	if (ct.name == "add") {
		auto p1 = generateI(ct.sub_tokens[0], e);
		auto p2 = generateI(ct.sub_tokens[1], e);
		if (ct.t->id == '+') return p2 + PUSH_EAX + p1 + POP_EDX + ADD_EAX_EDX;
		if (ct.t->id == '-') return p2 + PUSH_EAX + p1 + POP_EDX + SUB_EAX_EDX;
		throw "Unknown add OP "s + (char)ct.t->id;
	}
	if (ct.name == "mul") {
		auto p1 = generateI(ct.sub_tokens[0], e);
		auto p2 = generateI(ct.sub_tokens[1], e);
		if (ct.t->id == '*') return p2 + PUSH_EAX + p1 + POP_EDX + MUL_EAX_EDX;
		if (ct.t->id == '/')
			// TODO: properly handel sign extention
			return ZERO_EAX + PUSH_EAX + p2 + PUSH_EAX + p1 + POP_ECX + POP_EDX +
			       DIV_EDAX_ECX;
		if (ct.t->text == "mod")
			return ZERO_EAX + PUSH_EAX + p2 + PUSH_EAX + p1 + POP_ECX + POP_EDX +
			       MOD_EDAX_ECX;
		throw "Unknown mul OP "s + (char)ct.t->id;
	}
	if (ct.name == "exp") {
		if (ct.t->id == '^') return generateI(ct.sub_tokens[0], e);
		// TODO: impliment this
		throw "Unknown exp OP "s + (char)ct.t->id;
	}
	if (ct.name == "unary") {
		auto p = generateI(ct.sub_tokens[0], e);
		if (ct.t->id == '+') return p;
		if (ct.t->id == '-') return p + PUSH_EAX + ZERO_EAX + POP_EDX + SUB_EAX_EDX;
		throw "Unknown Unary Op "s + (char)ct.t->id;
	}
	if (ct.name == "block") {
		return accumulate(
		    ct.sub_tokens.begin(), ct.sub_tokens.end(), program(),
		    [&e](program p, Code_Tree c) { return p + generateI(c, e); });
	}
	if (ct.name == "print") {
		auto ret = program();
		for (auto c : ct.sub_tokens)
			if (c.name == "print_i")
				ret += generateI(c.sub_tokens[0], e) + PRINT_EAX_I;
			else if (c.name == "print_s") {
				auto p = e->push(c.sub_tokens[0].t->text);
				ret += LOADLL_STR((long long int)p) + PRINT_STR;
			} else if (c.name == "print_b") {
				ret += generateI(c.sub_tokens[0], e) + PRINT_EAX_BOOL;
			} else
				throw "Unknown print mode "s + c.name;
		return ret;
	}
	if (ct.name == "Declaration") {
		e->cleanup.push_back(POP_EAX);
		e->syms.push_back(ct.sub_tokens[0].t->text);
		return PUSH_EAX;
	}
	if (ct.name == "Ident") {
		auto var = ct.t->text;
		return MOV_EAX_STACK(e->lookup(var));
	}
	if (ct.name == "Read") {
		auto var = ct.sub_tokens[0].t->text;
		return READ_EAX + MOV_STACK_EAX(e->lookup(var));
	}
	if (ct.name == "Assignment") {
		auto var = ct.sub_tokens[0].t->text;
		return generateI(ct.sub_tokens[1], e) + MOV_STACK_EAX(e->lookup(var));
	}
	if (ct.name == "bool_rel") {
		auto ret = generateI(ct.sub_tokens[2], e) + PUSH_EAX +
		           generateI(ct.sub_tokens[1], e) + POP_ECX + CMP_EAX_ECX;
		if (ct.sub_tokens[0].t->id == '<')
			return ret + BOOL_IS_LESS;
		else if (ct.sub_tokens[0].t->id == LESS_EQ)
			return ret + BOOL_IS_LESS_EQ;
		else if (ct.sub_tokens[0].t->id == '=')
			return ret + BOOL_IS_EQ;
		else if (ct.sub_tokens[0].t->id == GREATER_EQ)
			return ret + BOOL_IS_GREATER_EQ;
		else if (ct.sub_tokens[0].t->id == '>')
			return ret + BOOL_IS_GREATER;
		else if (ct.sub_tokens[0].t->id == NOT_EQUAL)
			return ret + BOOL_IS_NE;
		else
			throw "Expected a boolean realational, got '"s +
			    ct.sub_tokens[0].t->text + "'";
	}
	if (ct.name == "bool_or") {
		return generateI(ct.sub_tokens[1], e) + PUSH_EAX +
		       generateI(ct.sub_tokens[0], e) + POP_ECX + BOOL_EAX_OR_ECX;
	}
	if (ct.name == "bool_and") {
		return generateI(ct.sub_tokens[1], e) + PUSH_EAX +
		       generateI(ct.sub_tokens[0], e) + POP_ECX + BOOL_EAX_AND_ECX;
	}
	if (ct.name == "bool_not") {
		return generateI(ct.sub_tokens[0], e) + BOOL_NOT;
	}
	throw "Bad Parse Tree - unknown '"s + ct.name + "'";
}
#define RETURN ((char)(0xC3))
program generate(Code_Tree ct) {
	program r;
	r.init();
	r += generateI(ct, r.e);
	r.cleanup();
	return r + RETURN;
}
