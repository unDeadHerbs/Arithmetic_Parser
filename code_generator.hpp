#ifndef __CODE_GENERATOR_HPP__
#define __CODE_GENERATOR_HPP__
#include "code_tree.hpp"
#include <utility>
#include <string.h>
#include <memory>

typedef unsigned char binary;

struct sym_props{};
struct program;
struct Env{
  char string_table[5000];
  char* end;
  std::vector<program> cleanup;
  std::vector<std::pair<std::string,sym_props>> syms;
public:
  Env():string_table{0},end(string_table){}
  char* push(std::string s){
    strcpy(end,s.c_str());
    auto ret=end;
    while(*end++);
    return ret;
  }
};

struct program{
  std::vector<char> buffer;
  std::shared_ptr<Env> e;
public:
  program(){}
  program(std::vector<char>b):buffer(b){}
  void operator()(); //TODO: add optional parameters
  void init(){e=std::make_shared<Env>();}
  void cleanup(){for(auto &c:e->cleanup)operator+=(c);}
  int size(){return buffer.size();}
  program& operator+=(program p){
    for(auto c:p.buffer)
      buffer.push_back(c);
    return *this;
  }
  program& operator+=(char c){
    buffer.push_back(c);
    return *this;
  }
  program operator+(program p){
    program t(*this);
    return t+=p;
  }
  program operator+(char c){
    program t(*this);
    return t+=c;
  }
  //operator += vec<opcode>
  //operator [] for indexing
  //Add an insert? Nope that's hard.
};
program generate(Code_Tree);

#endif
