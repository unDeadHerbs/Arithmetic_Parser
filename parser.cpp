#include "parser.hpp"

Code_Tree parse_expression(std::vector<Token>& tokens);

#define EOF_ERROR(x) Code_Tree(Token(ERROR, -1, -1, x))
#define ERROR(x, t) \
	Code_Tree("ERROR", Token(ERROR, t.line, t.col, x), {Code_Tree(t)})
#define EAT_OP(t, tokens, op)                                                \
	if (!tokens.size()) throw EOF_ERROR(std::string("Expected '") + op + "'"); \
	auto t = tokens[0];                                                        \
	if (t.id != op) throw ERROR(std::string("Expected '") + op + "'", t);      \
	tokens.erase(tokens.begin(), tokens.begin() + 1);

#define EAT_IDENT(t, tokens, str)                                             \
	if (!tokens.size()) throw EOF_ERROR(std::string("Expected '") + str + "'"); \
	auto t = tokens[0];                                                         \
	if (t.id != IDENT) throw ERROR(std::string("Expected '") + str + "'", t);   \
	if (t.text != str) throw ERROR(std::string("Expected '") + str + "'", t);   \
	tokens.erase(tokens.begin(), tokens.begin() + 1);

Code_Tree parse_literal(std::vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto t = tokens[0];
	tokens.erase(tokens.begin(), tokens.begin() + 1);
	if (t.id == INT || t.id == STRING || t.id == REAL)
		return Code_Tree("literal", t);
	throw ERROR("Not a Litteral", t);
}

Code_Tree parse_value(std::vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto t = tokens[0];
	if (t.id != '(') return parse_literal(tokens);
	tokens.erase(tokens.begin(), tokens.begin() + 1);
	auto ret = parse_expression(tokens);
	if (!tokens.size()) throw EOF_ERROR("Expected ')'.");
	t = tokens[0];
	if (t.id != ')') throw ERROR("Expected ')'.", t);
	tokens.erase(tokens.begin(), tokens.begin() + 1);
	return ret;
}

Code_Tree parse_unary_signs(std::vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto t = tokens[0];
	if (t.id != '+' && t.id != '-') return parse_value(tokens);
	tokens.erase(tokens.begin(), tokens.begin() + 1);
	return Code_Tree("unary", t, {parse_value(tokens)});
}

Code_Tree parse_exp(std::vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto ret = parse_unary_signs(tokens);
	if (!tokens.size()) return ret;
	auto t = tokens[0];
	if (t.id != '^') return ret;
	tokens.erase(tokens.begin(), tokens.begin() + 1);
	return Code_Tree("exp", t, {ret, parse_exp(tokens)});
}

Code_Tree parse_mul_div_prime(std::vector<Token>& tokens,
                              Code_Tree accumulator) {
	if (!tokens.size()) return accumulator;
	auto t = tokens[0];
	if (!(t.id == '*' || t.id == '/' || t.text == "mod")) return accumulator;
	tokens.erase(tokens.begin(), tokens.begin() + 1);
	return parse_mul_div_prime(
	    tokens, Code_Tree("mul", t, {accumulator, parse_exp(tokens)}));
}

Code_Tree parse_mul_div(std::vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto ret = parse_exp(tokens);
	return parse_mul_div_prime(tokens, ret);
}

Code_Tree parse_sum_dif_prime(std::vector<Token>& tokens,
                              Code_Tree accumulator) {
	if (!tokens.size()) return accumulator;
	auto t = tokens[0];
	if (!(t.id == '+' || t.id == '-')) return accumulator;
	tokens.erase(tokens.begin(), tokens.begin() + 1);
	return parse_sum_dif_prime(
	    tokens, Code_Tree("add", t, {accumulator, parse_mul_div(tokens)}));
}

Code_Tree parse_sum_dif(std::vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	auto ret = parse_mul_div(tokens);
	return parse_sum_dif_prime(tokens, ret);
}

Code_Tree parse_expression(std::vector<Token>& tokens) {
	return parse_sum_dif(tokens);
}

Code_Tree parse_printable(std::vector<Token>& tokens) {
	return parse_expression(tokens);
	// else parse string
}

Code_Tree parse_printables_prime(std::vector<Token>& tokens,
                                 Code_Tree accumulator) {
	try {
		EAT_OP(t, tokens, ',');
		return parse_printables_prime(
		    tokens, Code_Tree("print", t, {accumulator, parse_printable(tokens)}));

	} catch (Code_Tree err) {
		return accumulator;
	}
}
Code_Tree parse_printables(std::vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROR");
	return parse_printables_prime(tokens, parse_printable(tokens));
}

Code_Tree parse_print_f(std::vector<Token>& tokens) {
	EAT_IDENT(t1, tokens, "print");
	EAT_OP(t2, tokens, '(');
	auto ret = parse_printables(tokens);
	EAT_OP(t3, tokens, ')');
	EAT_OP(t4, tokens, ';');
	return Code_Tree("print", t1, {ret});
}

Code_Tree parse_statement(std::vector<Token>& tokens) {
	return parse_print_f(tokens);
}

Code_Tree parse_statements(std::vector<Token>& tokens, Code_Tree accumulator) {
	if (!tokens.size()) return accumulator;
	auto t = tokens[0];
	if (t.id == '}') return accumulator;
	return parse_statements(
	    tokens, Code_Tree("block", t, {accumulator, parse_statement(tokens)}));
}

Code_Tree parse_global_block(std::vector<Token>& tokens) {
	if (!tokens.size()) throw EOF_ERROR("EOF ERROF");
	return Code_Tree("block",
	                 Token(TOKEN('{'), tokens[0].line, tokens[0].col, "block"),
	                 {parse_statements(tokens, parse_statement(tokens))});
}
Code_Tree parse_block(std::vector<Token>& tokens) {
	EAT_OP(t1, tokens, '{');
	auto ret = parse_global_block(tokens);
	EAT_OP(t2, tokens, '}');
	return ret;
}

Code_Tree parse(std::vector<Token>& tokens) {
	try {
		return parse_global_block(tokens);
	} catch (Code_Tree err) {
		return err;
	}
}
