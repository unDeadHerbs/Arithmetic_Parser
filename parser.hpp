#ifndef __PARSER_HPP__
#define __PARSER_HPP__

#include <string>
#include <variant>
#include <vector>
#include "Tokenizer/tokenizer.hpp"
#include "code_tree.hpp"

Code_Tree parse(std::vector<Token>&);

#endif
