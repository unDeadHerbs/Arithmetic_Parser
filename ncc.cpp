#include <iostream>
using std::cout;
using std::endl;
#include "Tokenizer/tokenizer.hpp"
#include "code_generator.hpp"
#include "parser.hpp"
using std::ostream;
using std::string;
#include <iomanip>
using std::setw;
#include <vector>
using std::vector;

int main(int argc, char* argv[]) {
	vector<Token> tokens;
	if (argc != 2) {
		cout << "Usage: " << argv[0] << " [Filename]" << endl;
		return 0;
	}
	if (!initialize(argv[1])) exit(-1);
	while (Token t = get_token()) tokens.push_back(t);
	cleanup();
	//*
	cout << "| Token Type | Token Value" << endl
	     << "+------------+-- - - -  -  -" << endl;
	for (auto& t : tokens) cout << t << endl;
	cout << "+------------+-- - - -  -  -" << endl
	     << endl
	     << endl
	     << "Parsing" << endl;  //*/
	while (tokens.size()) {
		auto ct = parse(tokens);
		cout << "Code tree: " << endl << ct;
		// if (ct.name == "Error") continue;
		// auto prog = generate(ct);
		cout << "Code size: " << /*prog.size()<<*/ " bytes." << endl
		     << "Code execution:" << endl /*<<prog()<<endl*/;
	}
	cout.flush();
}
