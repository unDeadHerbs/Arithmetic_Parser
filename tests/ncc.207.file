# print statement

print( "Hello,", " World!" );
