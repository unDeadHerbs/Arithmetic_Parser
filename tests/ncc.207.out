Code tree: 
block
  print
    Hello,
     World!
Code size: 49 bytes.
Code execution:
Hello, World!

